executor/src/subscriber.rs

```diff
@@ -1,95 +1,103 @@
 pub mod subscriber_tests;
 
 /// The `Subscriber` receives certificates sequenced by the consensus and waits until the
-/// `BatchLoader` downloaded all the transactions references by the certificates; it then
+/// downloaded all the transactions references by the certificates; it then
 /// forward the certificates to the Executor Core.
 pub struct Subscriber {
     /// The temporary storage holding all transactions' data (that may be too big to hold in memory).
-    store: Store<BatchDigest, SerializedBatchMessage>,
+    store: Store<BatchDigest, Batch>,
     /// Receive reconfiguration updates.
     rx_reconfigure: watch::Receiver<ReconfigureNotification>,
     /// A channel to receive consensus messages.
     rx_consensus: metered_channel::Receiver<ConsensusOutput>,
-    /// A channel to the batch loader to download transaction's data.
-    tx_batch_loader: metered_channel::Sender<ConsensusOutput>,
     /// A channel to send the complete and ordered list of consensus outputs to the executor. This
     /// channel is used once all transactions data are downloaded.
     tx_executor: metered_channel::Sender<ConsensusOutput>,
+    // A channel to send commands to the block waiter to receive
+    // a certificate's batches (block).
+    tx_get_block_commands: metered_channel::Sender<BlockCommand>,
+    // When asking for a certificate's payload we want to retry until we succeed, unless
+    // some irrecoverable error occurs. For that reason a backoff policy is defined
+    get_block_retry_policy: ExponentialBackoff,
 }
 
 impl Subscriber {
     /// Spawn a new subscriber in a new tokio task.
     #[must_use]
     pub fn spawn(
-        store: Store<BatchDigest, SerializedBatchMessage>,
+        store: Store<BatchDigest, Batch>,
+        tx_get_block_commands: metered_channel::Sender<BlockCommand>,
         rx_reconfigure: watch::Receiver<ReconfigureNotification>,
         rx_consensus: metered_channel::Receiver<ConsensusOutput>,
-        tx_batch_loader: metered_channel::Sender<ConsensusOutput>,
         tx_executor: metered_channel::Sender<ConsensusOutput>,
     ) -> JoinHandle<()> {
+        let get_block_retry_policy = ExponentialBackoff {
+            initial_interval: Duration::from_millis(500),
+            randomization_factor: backoff::default::RANDOMIZATION_FACTOR,
+            multiplier: backoff::default::MULTIPLIER,
+            max_interval: Duration::from_secs(10), // Maximum backoff is 10 seconds
+            max_elapsed_time: None, // Never end retrying unless a non recoverable error occurs.
+            ..Default::default()
+        };
+
         tokio::spawn(async move {
             Self {
                 store,
                 rx_reconfigure,
                 rx_consensus,
-                tx_batch_loader,
                 tx_executor,
+                tx_get_block_commands,
+                get_block_retry_policy,
             }
             .run()
             .await
             .expect("Failed to run subscriber")
```

Comment:
ha cool, if an unrecoverable error happens we shut down here.

Reply:
Exactly, please also see the test [here](https://github.com/MystenLabs/narwhal/pull/738/files#diff-ff908d558501ea2dad76cf598def17b4e8fa49c04205323a81f896d81fd4cc42R174) where I am explicitly checking that subscriber will crash in this case.

You can also see the recoverable / irrecoverable cases [here](https://github.com/MystenLabs/narwhal/pull/738/files#diff-dd570646b526219ac76c68c9546526c43df1bf27085a0949d1355a15bd5d84b8R129) . Errors that are mapped to `Transient` errors will make the retrier retry again. Errors that are marked as `Permanent` will make the retrier stop and eventually make the subscriber crash.

Reply:
https://github.com/MystenLabs/narwhal/pull/707 should also help with the crashing :)